packages/dev/core/src/Rendering/GaussianSplatting/gaussianSplatting.ts

```diff
@@ -0,0 +1,342 @@
+     */
+    public loadAsync(url: string): Promise<void> {
+        return Tools.LoadFileAsync(url, true).then((data: string | ArrayBuffer) => {
+            if (this.mesh) {
+                this.dispose();
+            }
+            this._setData(new Uint8Array(data as any));
+            const matricesData = new Float32Array(this.vertexCount * 16);
+
+            const updateInstances = (indexMix: Uint32Array) => {
+                for (let j = 0; j < this.vertexCount; j++) {
+                    const i = indexMix[2 * j];
+                    const index = j * 16;
+                    matricesData[index + 0] = this._positions[i * 3 + 0];
+                    matricesData[index + 1] = this._positions[i * 3 + 1];
+                    matricesData[index + 2] = this._positions[i * 3 + 2];
+
+                    matricesData[index + 4] = this._uBuffer[32 * i + 24 + 0] / 255;
+                    matricesData[index + 5] = this._uBuffer[32 * i + 24 + 1] / 255;
+                    matricesData[index + 6] = this._uBuffer[32 * i + 24 + 2] / 255;
+                    matricesData[index + 7] = this._uBuffer[32 * i + 24 + 3] / 255;
+
+                    matricesData[index + 8] = this._covA[i * 3 + 0];
+                    matricesData[index + 9] = this._covA[i * 3 + 1];
+                    matricesData[index + 10] = this._covA[i * 3 + 2];
+
+                    matricesData[index + 12] = this._covB[i * 3 + 0];
+                    matricesData[index + 13] = this._covB[i * 3 + 1];
+                    matricesData[index + 14] = this._covB[i * 3 + 2];
+                }
+
+                this.mesh?.thinInstanceBufferUpdated("matrix");
+            };
+
+            // update so this.mesh is valid when exiting this function
+            this.mesh = this._getMesh(this.scene);
+            this.mesh.thinInstanceSetBuffer("matrix", matricesData, 16, false);
+
+            if (GaussianSplatting._Worker) {
+                GaussianSplatting._Worker.terminate();
+            }
+
+            GaussianSplatting._Worker = new Worker(
+                URL.createObjectURL(
+                    new Blob(["(", GaussianSplatting._CreateWorker.toString(), ")(self)"], {
+                        type: "application/javascript",
+                    })
+                )
+            );
+
+            GaussianSplatting._Worker.onmessage = (e) => {
+                const indexMix = new Uint32Array(e.data.depthMix.buffer);
+                updateInstances(indexMix);
+            };
+            this._sceneBeforeRenderObserver = this.scene.onBeforeRenderObservable.add(() => {
+                const engine = this.scene.getEngine();
+                this._material.setVector2("viewport", new Vector2(engine.getRenderWidth(), engine.getRenderHeight()));
+                this.mesh!.getWorldMatrix().multiplyToRef(this.scene.activeCamera!.getViewMatrix(), TmpVectors.Matrix[0]);
+                this._material.setMatrix("modelView", TmpVectors.Matrix[0]);

```

Comment:
That's a bit dangerous to pass a tmp matrix to `setMatrix`, because this method does not copy the matrix, it takes it by reference. So, if before the mesh is rendered, some code elsewhere modifies `TmpVectors.Matrix[0]`, the matrix won't be the right one. Maybe you can use a temp matrix local to the class?

Reply:
fixed with a class member